jetty-http/src/main/java/org/eclipse/jetty/http/compression/Huffman.java

```diff
@@ -348,199 +347,46 @@
-        int len = s.length();
-        for (int i = 0; i < len; i++)
-        {
-            char c = s.charAt(i);
-            if (c >= 128 || c < ' ')
-                return -1;
-            needed += table[c][1];
-        }
-
-        return (needed + 7) / 8;
-    }
-
-    private static int octetsNeeded(final int[][] table, byte[] b)
+    public static boolean isIllegalCharacter(char c)
     {
-        int needed = 0;
-        int len = b.length;
-        for (int i = 0; i < len; i++)
-        {
-            int c = 0xFF & b[i];
-            needed += table[c][1];
-        }
-        return (needed + 7) / 8;
+        return (c >= 256 || c < ' ');
     }
 
-    /**
-     * @param table The table to encode by
-     * @param buffer The buffer to encode to
-     * @param s The string to encode
-     */
-    private static void encode(final int[][] table, ByteBuffer buffer, String s)
+    public static char getReplacementCharacter(char c)
     {
-        long current = 0;
-        int n = 0;
-        int len = s.length();
-        for (int i = 0; i < len; i++)
+        switch (c)
         {
-            char c = s.charAt(i);
-            if (c >= 128 || c < ' ')
-                throw new IllegalArgumentException();
-            int code = table[c][0];
-            int bits = table[c][1];
-
-            current <<= bits;
-            current |= code;
-            n += bits;
-
-            while (n >= 8)
-            {
-                n -= 8;
-                buffer.put((byte)(current >> n));
-            }
+            // A recipient of CR, LF, or NUL within a field value MUST either reject the message
+            // or replace each of those characters with SP before further processing
+            case '\r':
+            case '\n':
+            case 0x00:
+                return ' ';
+
+            default:
+                if (c >= 256 || c < ' ')
+                    return '?';
         }
 
-        if (n > 0)
-        {
-            current <<= (8 - n);
-            current |= (0xFF >>> n);
-            buffer.put((byte)(current));
-        }
+        return c;
     }
 
-    private static void encode(final int[][] table, ByteBuffer buffer, byte[] b)
+    public static int getReplacementCharacter(int i)
     {
-        long current = 0;
-        int n = 0;
-
-        int len = b.length;
-        for (int i = 0; i < len; i++)
+        switch (i)
         {
-            int c = 0xFF & b[i];
-            int code = table[c][0];
-            int bits = table[c][1];
-
-            current <<= bits;
-            current |= code;
-            n += bits;
-
-            while (n >= 8)
-            {
-                n -= 8;
-                buffer.put((byte)(current >> n));
-            }
+            // A recipient of CR, LF, or NUL within a field value MUST either reject the message
+            // or replace each of those characters with SP before further processing
+            case '\r':
+            case '\n':
+            case 0x00:
```

Comment:
this assumes a ASCII compatible charset, so why not just cast to a char and call the other method?